### ISSUE TITLE
commitlog_replayer - Corrupted file: .... after restart

### ISSUE THREAD

**Issue by ShlomiBalalis** [scylladb/scylladb#4700]
- [x] I have read the disclaimer above, and I am reporting a suspected malfunction in Scylla.

*Installation details*
Scylla version (or git commit hash):3.1.0.rc2-0.20190630.acff367ea
Cluster size:4
OS (RHEL/CentOS/Ubuntu/AWS AMI):AWS AMI

After restarting a one of the nodes, some files appeared to be corrupted
Also, for an unknown reason there was a 7 minute gap in the log at the same time

[journal_log.txt](https://github.com/scylladb/scylla/files/3381782/journal_log.txt)

**Comment by tgrabiec** [comment]
Also observed here: https://github.com/scylladb/scylla/issues/4647#issuecomment-509884135

**Comment by slivne** [comment]
```
commitlog_replayer - Corrupted file: /var/lib/scylla/commitlog/CommitLog-1-162129586585742354.log. 16252920 bytes skipped.
```

**Comment by slivne** [comment]
@elcallio 

- I think this maybe expected in case we crash a server  (sigkill)
- I do not think this is expected if the server is stopped (in this case it was stopped)

**Comment by slivne** [comment]
@elcallio ping

**Comment by elcallio** [comment]
If the server was stopped, there should be no commitlog files at all. If it was killed -9, having some files corrupted is certainly possible, but this seem like to many to be sane. That being said, it looks like it is one segment per shard that is borked. 

Now it is possible that this is also wholly (semi-)benign, with these segments being reused segments that where not terminated properly (because of the whole kill -9 thing). This will cause the reader to eventually think "corrupted" when he gets to the "old" part of the file. This should not be a big issue though, except for some ugly logging. It just means we terminated replay without a null-page mark. 

From the data here I cannot say either way for sure. Can we repro this? I assume this was an AWS run? 
Setting commitlog logging to debug would probably shed some more light.

**Comment by slivne** [comment]
there was as an attempt to shutdown the server cleanly yet we timeout in the clean shutdown

```
Jul 10 21:25:40 ip-10-0-229-245.eu-west-1.compute.internal systemd[1]: Stopping Scylla Server...
Jul 10 21:25:40 ip-10-0-229-245.eu-west-1.compute.internal scylla.bin[23501]:  [shard 0] init - Signal received; shutting down
Jul 10 21:25:40 ip-10-0-229-245.eu-west-1.compute.internal scylla.bin[23501]:  [shard 0] compaction_manager - Asked to stop
Jul 10 21:25:40 ip-10-0-229-245.eu-west-1.compute.internal scylla.bin[23501]:  [shard 5] compaction_manager - Asked to stop
.
.
.
Jul 10 21:25:40 ip-10-0-229-245.eu-west-1.compute.internal scylla.bin[23501]:  [shard 9] compaction_manager - compaction info: Compaction for keyspace1/standard1 was stopped due to shutdown.: will stop
Jul 10 21:25:40 ip-10-0-229-245.eu-west-1.compute.internal scylla.bin[23501]:  [shard 9] compaction_manager - Stopped
Jul 10 21:25:40 ip-10-0-229-245.eu-west-1.compute.internal scylla.bin[23501]:  [shard 0] compaction_manager - Stopped
Jul 10 21:25:40 ip-10-0-229-245.eu-west-1.compute.internal scylla.bin[23501]:  [shard 5] compaction_manager - Stopped
Jul 10 21:25:40 ip-10-0-229-245.eu-west-1.compute.internal scylla.bin[23501]:  [shard 10] compaction_manager - compaction info: Compaction for keyspace_deflate/standard1 was stopped due to shutdown.: will stop
Jul 10 21:25:40 ip-10-0-229-245.eu-west-1.compute.internal scylla.bin[23501]:  [shard 1] compaction_manager - compaction info: Compaction for keyspace1/standard1 was stopped due to shutdown.: will stop
Jul 10 21:25:40 ip-10-0-229-245.eu-west-1.compute.internal scylla.bin[23501]:  [shard 10] compaction_manager - Stopped
Jul 10 21:25:40 ip-10-0-229-245.eu-west-1.compute.internal scylla.bin[23501]:  [shard 1] compaction_manager - Stopped
Jul 10 21:25:40 ip-10-0-229-245.eu-west-1.compute.internal scylla.bin[23501]:  [shard 0] view - Stopping view builder
Jul 10 21:25:40 ip-10-0-229-245.eu-west-1.compute.internal scylla.bin[23501]:  [shard 13] view - Stopping view builder
Jul 10 21:25:40 ip-10-0-229-245.eu-west-1.compute.internal scylla.bin[23501]:  [shard 11] view - Stopping view builder
Jul 10 21:25:40 ip-10-0-229-245.eu-west-1.compute.internal scylla.bin[23501]:  [shard 6] view - Stopping view builder
Jul 10 21:25:40 ip-10-0-229-245.eu-west-1.compute.internal scylla.bin[23501]:  [shard 8] view - Stopping view builder
Jul 10 21:25:40 ip-10-0-229-245.eu-west-1.compute.internal scylla.bin[23501]:  [shard 3] view - Stopping view builder
Jul 10 21:25:40 ip-10-0-229-245.eu-west-1.compute.internal scylla.bin[23501]:  [shard 12] view - Stopping view builder
Jul 10 21:25:40 ip-10-0-229-245.eu-west-1.compute.internal scylla.bin[23501]:  [shard 4] view - Stopping view builder
Jul 10 21:25:40 ip-10-0-229-245.eu-west-1.compute.internal scylla.bin[23501]:  [shard 1] view - Stopping view builder
Jul 10 21:25:40 ip-10-0-229-245.eu-west-1.compute.internal scylla.bin[23501]:  [shard 5] view - Stopping view builder
Jul 10 21:25:40 ip-10-0-229-245.eu-west-1.compute.internal scylla.bin[23501]:  [shard 7] view - Stopping view builder
Jul 10 21:25:40 ip-10-0-229-245.eu-west-1.compute.internal scylla.bin[23501]:  [shard 2] view - Stopping view builder
Jul 10 21:25:40 ip-10-0-229-245.eu-west-1.compute.internal scylla.bin[23501]:  [shard 10] view - Stopping view builder
Jul 10 21:25:40 ip-10-0-229-245.eu-west-1.compute.internal scylla.bin[23501]:  [shard 9] view - Stopping view builder
Jul 10 21:25:40 ip-10-0-229-245.eu-west-1.compute.internal scylla.bin[23501]:  [shard 0] storage_service - Drain on shutdown: starts
Jul 10 21:25:40 ip-10-0-229-245.eu-west-1.compute.internal scylla.bin[23501]:  [shard 0] storage_service - Stop transport: starts
Jul 10 21:25:40 ip-10-0-229-245.eu-west-1.compute.internal scylla.bin[23501]:  [shard 0] storage_service - Thrift server stopped
Jul 10 21:27:10 ip-10-0-229-245.eu-west-1.compute.internal systemd[1]: scylla-server.service stop-sigterm timed out. Killing.
Jul 10 21:27:11 ip-10-0-229-245.eu-west-1.compute.internal systemd[1]: scylla-server.service: main process exited, code=killed, status=9/KILL
Jul 10 21:27:11 ip-10-0-229-245.eu-west-1.compute.internal systemd[1]: Stopped Scylla Server.
Jul 10 21:27:11 ip-10-0-229-245.eu-west-1.compute.internal systemd[1]: Unit scylla-server.service entered failed state.
Jul 10 21:27:11 ip-10-0-229-245.eu-west-1.compute.internal systemd[1]: scylla-server.service failed.
Jul 10 21:27:17 ip-10-0-229-245.eu-west-1.compute.internal systemd[1]: Starting Scylla Server...
.
.
ul 10 21:27:48 ip-10-0-229-245.eu-west-1.compute.internal scylla.bin[54723]:  [shard 9] commitlog_replayer - Corrupted file: /var/lib/scylla/commitlog/CommitLog-1-162129586585742354.log. 16252920 bytes skipped.
Jul 10 21:27:48 ip-10-0-229-245.eu-west-1.compute.internal scylla.bin[54723]:  [shard 7] commitlog_replayer - Corrupted file: /var/lib/scylla/commitlog/CommitLog-1-126100789566778314.log. 14680056 bytes skipped.
Jul 10 21:27:48 ip-10-0-229-245.eu-west-1.compute.internal scylla.bin[54723]:  [shard 13] commitlog_replayer - Corrupted file: /var/lib/scylla/commitlog/CommitLog-1-234187180623670211.log. 15380472 bytes skipped.
```

not clear to me if we are expecting scylla to shutdown cleanly in less than 3 minutes or not

So it seems the DefaultTimeoutStopSec is 90s, 

We change the boot setting to be TimeoutStartSec=900

We should probably adjust the shutdown as well 

@avikivity 

which means setting the TimeoutStopSec=900 as well

**Comment by tgrabiec** [comment]
I think that kill -9 should not result in corruption of the commit log. The reason we have a commit log is to preserve writes in case of a crash.

Printing "corrupted" when the segments are not really corrupted is not very nice because it suggests data loss when there is none. So we actually won't trust it, so it carries no information, and actual data loss cannot be detected.

**Comment by elcallio** [comment]
The problem is that short of pre-writing zeros in the whole re-used segment, there is not really any safe way to tell the difference between a corrupt chunk and one that was partially overwritten by non-terminated re-use.

**Comment by tgrabiec** [comment]
Maybe we could assign some sort of generation number and prefix each block with it a segment with it. If we see the generation number going down, it means we crossed overwrite boundary.

**Comment by tgrabiec** [comment]
How do we currently detect that we crossed overwrite boundary?

**Comment by elcallio** [comment]
Yes, that was my initial idea also. Problem is that 
a.) Chunks are of different size, so we will not necessarily read the sequence number when crossing relevant/remaining boundary. Ok, this is sort of fine, because this probably means we'll read the wrong value anyway. 
but:
b.) Reading a value that is not correct is technically indistinguishable from a corrupted chunk. I.e. the normal chunk crc check that fails and generates the "corrupt segment" warning in the first place is really just a different expression of the same check. So back to square one.

**Comment by elcallio** [comment]
If a segment is properly terminated, we write a set of zeroes in place of the chunk header as an end marker.

**Comment by tgrabiec** [comment]
Could we make chunks fixed-size? Large writes would have to be split into
several chunks. That could be hidden behind some stream-like abstraction
which offers a flat space.

**Comment by elcallio** [comment]
We could, and include present segment id in each header, and use this as end marker instead. But the problem is that we always need to (or at least should) write the full mutation in a write-flush cycle. This would require us to write partial chunks if we cross the boundary (or accept even more slack than we have now), which sort of would re-create the exact same problem as now - how do you know if a chunk is complete? Basically the same problem but with data entry header instead.

**Comment by tgrabiec** [comment]
On Tue, Jul 16, 2019 at 6:51 PM Calle Wilund <notifications@github.com>
wrote:

> We could, and include present segment id in each header, and use this as
> end marker instead. But the problem is that we always need to (or at least
> should) write the full mutation in a write-flush cycle. This would require
> us to write partial chunks if we cross the boundary (or accept even more
> slack than we have now), which sort of would re-create the exact same
> problem as now - how do you know if a chunk is complete? Basically the same
> problem but with data entry header instead.
>
When flushing, we can append an end-marker (e.g. data entry with zero
size), which will be overwritten if we append more.

**Comment by elcallio** [comment]
No, not really, because we are allowed to write said are (that is being zeroed) while the flush is ongoing, so we would potentially be writing the same block twice simultaneously. This might be ok, assuming we get the ordering of continuations right, and the fs/disk orders the writes, but I am still scared of it...

**Comment by avikivity** [comment]
We should make commitlog self-synchronizing so we can tell the difference between real corruption and ordinary segment reuse.

**Comment by avikivity** [comment]
> On Tue, Jul 16, 2019 at 6:51 PM Calle Wilund ***@***.***> wrote: We could, and include present segment id in each header, and use this as end marker instead. But the problem is that we always need to (or at least should) write the full mutation in a write-flush cycle. This would require us to write partial chunks if we cross the boundary (or accept even more slack than we have now), which sort of would re-create the exact same problem as now - how do you know if a chunk is complete? Basically the same problem but with data entry header instead.
> When flushing, we can append an end-marker (e.g. data entry with zero size), which will be overwritten if we append more.

This will cause a lot of amplification, particularly if we drop the 10-second periodic mode in favor of batch mode with O_DSYNC. I think each commilog block (say 4k) should be in a frame with a generation number like you propose and its own CRC.

**Comment by elcallio** [comment]
Every block (chunk) does have its own CRC, however, we only tag the segment header with segment ID (thus if you try to replay a recycled segment, the header ID and the filename ID will mismatch, and we skip it.
But if we crash before terminating a recycled block, we will potentially hit a "chunk" that is just garbage left over from previous usage. This will fail CRC checks (most likely) so we say "corrupted data" and stop. 
Problem is that unless we switch to fixed chunk sizes, adding a segment ID to chunks here would not help, since there is no way to distinguish hitting "old data" from corrupted file parts. 
Of course, doing a proper pre-zeroing of recycled segments would fix this, but adds a lot of IO...

**Comment by avikivity** [comment]
We could have this:

1. A unique marker for every generation of a file. Could be an updated name, could be some data at the beginning of the file.
2. Every chunk has a copy of the generation number, plus the CRC
3. Every chunk has a signature that is unlikely to happen in regular data

This way we can distinguish between an old chunk (will have wrong generation) and a corrupted chunk (correct generation but bad CRC).

Though I think moving to fixed-size chunks is better.

**Comment by avikivity** [comment]
We can also consider a cryptographic hash, I think the performance should be okay.

**Comment by elcallio** [comment]
Unless chunks are fixed-size it does not help. Because you cannot distinguish hitting random left-over data where next chunk is expected to start from actual disk file corruption. Again, problem is that we hit stuff that is _not_ a chunk.

**Comment by avikivity** [comment]
That's why I wanted a signature. Of course, it doesn't prevent the problem completely because the signature can appear in random data, but with a 16-byte signature that's unlikely.

But yes, fixed-size blocks are better.

**Comment by elcallio** [comment]
Since the issue was that we say "corrupted" when we are actually just hitting the end of newly written data and encounter recycled data instead, there is no way that a signature will change this the only thing we will know is that we are not seeing a (valid) chunk. Fixed sized blocks would allow us to verify that a block is ok, but old, yes, but then we have to break data up between chunks, which _significantly_ changes both read and write logic. Of course, we don't need to be backwards compatible, but still...
The bigger issue is that fixed size would cause significant waste for batch-mode or forced disk write paths, since we cannot write _less_ than a chunk. We could of course continue a write buffer and rewrite to disk over and over until full, but then we are actually potentially loosing data in the overwrite.

**Comment by avikivity** [comment]
Item 1 was a unique generation number. So hitting an old block would see it referring to the wrong generation number.

**Comment by elcallio** [comment]
We already do write the segment ID into the chunk header. We can absolutely distinguish between an old and current chunk, should we ever hit a case where we manage to perfectly match into the old datas position. But that does not really happen. Problem is classifying garbage: Is it just recycled tail data or actual disk corruption.

**Comment by elcallio** [comment]
I realize I said above we did not tag chunks with ID. Sorry, I meant data entries.

**Comment by mykaul** [comment]
@elcallio - is this issue still relevant?

**Comment by avikivity** [comment]
Fixed by 60af2f3cb21914f855fd5cc1929629c113b7999b
